pkg/networkservice/xconnect/l3xconnect/common.go

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2021 Doc.ai and/or its affiliates.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package l3xconnect
+
+import (
+	"context"
+	"net"
+	"time"
+
+	"git.fd.io/govpp.git/api"
+	"github.com/edwarnicke/govpp/binapi/fib_types"
+	"github.com/edwarnicke/govpp/binapi/interface_types"
+	"github.com/edwarnicke/govpp/binapi/l3xc"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/api/pkg/api/networkservice"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/mechanisms/cls"
+	"github.com/networkservicemesh/api/pkg/api/networkservice/payload"
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+
+	"github.com/networkservicemesh/sdk-vpp/pkg/tools/ifindex"
+	"github.com/networkservicemesh/sdk-vpp/pkg/tools/types"
+)
+
+func createXc(swIfIndexFrom, swIfIndexTo uint32, nextHop net.IP) l3xc.L3xc {
+	isIP6 := nextHop.To4() == nil
+	return l3xc.L3xc{
+		SwIfIndex: interface_types.InterfaceIndex(swIfIndexFrom),
+		IsIP6:     isIP6,
+		NPaths:    1,
+		Paths: []fib_types.FibPath{
+			{
+				SwIfIndex: swIfIndexTo,
+				Weight:    1,
+				Type:      fib_types.FIB_API_PATH_TYPE_NORMAL,
+				Proto:     types.IsV6toFibProto(isIP6),
+				Nh: fib_types.FibPathNh{
+					Address: types.ToVppAddress(nextHop).Un,
+				},
+			},
+		},
+	}
+}
+
+func create(ctx context.Context, conn *networkservice.Connection, vppConn api.Connection) error {
+	// L3xc works only with IP payload
+	if conn.GetPayload() != payload.IP {
+		return nil
+	}
+
+	clientIfIndex, ok := ifindex.Load(ctx, true)
+	if !ok {
+		return nil
+	}
+	serverIfIndex, ok := ifindex.Load(ctx, false)
+	if !ok {
+		return nil
+	}
+
+	nextHopSrcIP := conn.GetContext().GetIpContext().GetSrcIPNet().IP
+	nextHopDstIP := conn.GetContext().GetIpContext().GetDstIPNet().IP
+	isDirectOrder := conn.GetMechanism().GetCls() == cls.LOCAL
+
+	// Set undefined index to allow vpp to choose an appropriate interface to remote by itself
```

Comment:
I don't quite follow why we are using undefined interfaces here... how does that work exactly?

Reply:
And why is it related to the cls.LOCAL?

Reply:
It is a good question. The cross connect for `Remote->Local` (wireguard to memif for example) always works correctly. But there is a problem with `Local->Remote`. I tested it many times with valid value of interface index, but for some reason vpp doesn't choose the route. It works only if we set undefined index and allow vpp to choose the right interface based on only nextHop.
Moreover, there is the same problem if we create `l3xc` manually using vppctl. It depends on the order of argument. If we run `l3xc add memif1/0 via wg0 172.16.1.100` - all works correctly. But if we change it to `l3xc add memif1/0 via 172.16.1.100 wg0` - it doesn't.

Reply:
How does VPP 'choose the right interface' ?

Reply:
Please note: we shouldn't be programming 'routes' into vpp because we have no way to prevent them colliding... the point of l3cx is *not* to have to program routes into the routing table.

Reply:
> If we run l3xc add memif1/0 via wg0 172.16.1.100 - all works correctly. But if we change it to l3xc add memif1/0 via 172.16.1.100 wg0 - it doesn't.

Why would the 'via' IP address be the same in both directions?

Reply:
> Why would the 'via' IP address be the same in both directions?

These examples for the the same direction. I just wanted to show that the order after "via" affects behavior.

Reply:
After wireguard creating we already have an entry in routing table. We don't create routes ourselves. So vpp should choose this route for l3xc. But for some reason this does not happen.

Reply:
OK... I've poked at this a bit more... I think this is what is happening.

the allowed-ips for the wireguard config are inserting routes into the fib (we really really really need to not do this, as there *will* be conflicts between connections).  When you set the via ip and *not* the interface, its using the ip fib to lookup the interface.  That's why its working.

However, we need to *not* have routes inserted into the fib by allowed-ips, and thus this will *not* work once that's corrected.

In poking at things by hand, you are correct that ```l3xc add memif1/0 via wg0 172.16.1.100``` works correctly... so does ```l3xc add memif1/0 via wg0``` ... which leads me to suspect that the ip address is being ignored in the  ```l3xc add memif1/0 via wg0 172.16.1.100``` and its just going via the interface.

Still poking at it a bit myself :)  Feel free to chime in with insights :)

Reply:
OK.. I've figured out that simply leaving the allowed-ips empty results in not having any of the routes in the ip fib.  Which is good.